source/val/validate_composites.cpp

```diff
@@ -618,8 +618,493 @@ spv_result_t ValidateCopyLogical(ValidationState_t& _,
+            n_rows == 32) &&
+          !(_.ContainsSizedIntOrFloatType(result_elt_type, spv::Op::OpTypeFloat,
+                                          16) &&
+            n_rows == 16) &&
+          !(_.ContainsSizedIntOrFloatType(result_elt_type, spv::Op::OpTypeFloat,
+                                          32) &&
+            n_rows == 8)) {
+        return _.diag(SPV_ERROR_INVALID_DATA, inst)
+               << "Opcode " << spvOpcodeString(inst->opcode())
+               << " requires the rows of the result operand have the bit "
+                  "length of 256"
+               << " when result coopmat's use is MatrixBKHR";
+      }
+      // source array length check
+      if (!is_src_arr_len_spec_const &&
+          !(_.ContainsSizedIntOrFloatType(source_elt_type, spv::Op::OpTypeInt,
+                                          32) &&
+            _.IsUnsignedIntScalarType(source_elt_type) && (ar_len == 8)) &&
+          !(n_rows == ar_len)) {
+        return _.diag(SPV_ERROR_INVALID_DATA, inst)
+               << "Opcode " << spvOpcodeString(inst->opcode())
+               << " requires the source array length be 8 if its elt type is "
+                  "32-bit unsigned OpTypeInt and be the result's number of "
+                  "rows, otherwise"
+               << " when result coopmat's use is MatrixBKHR";
+      }
+      break;
+    case spv::CooperativeMatrixUse::MatrixAccumulatorKHR:
+      // result coopmat component type check
+      if (!IsIntNOrFP32OrFP16<32>(_, result_elt_type)) {
+        return _.diag(SPV_ERROR_INVALID_DATA, inst)
+               << "Opcode " << spvOpcodeString(inst->opcode())
+               << " requires the result element type is one of 32-bit "
+                  "OpTypeInt signed/unsigned, 16- or 32-bit OpTypeFloat"
+               << " when result coopmat's use is MatrixAccumulatorKHR";
+      }
+      // source array length check
+      if (!is_res_col_spec_const && !is_src_arr_len_spec_const &&
+          !(_.ContainsSizedIntOrFloatType(source_elt_type, spv::Op::OpTypeInt,
+                                          32) &&
+            _.IsUnsignedIntScalarType(source_elt_type) &&
+            (_.ContainsSizedIntOrFloatType(result_elt_type,
+                                           spv::Op::OpTypeFloat, 16)
+                 ? (n_cols / 2 == ar_len)
+                 : n_cols == ar_len)) &&
+          (n_cols != ar_len)) {
+        return _.diag(SPV_ERROR_INVALID_DATA, inst)
+               << "Opcode " << spvOpcodeString(inst->opcode())
+               << " requires the source array length be a half of the number "
+                  "of columns of the resulting cooerative matrix if the "
+                  "matrix's componet type is 16-bit OpTypeFloat and be equal "
+                  "to the number of columns, otherwise,"
+               << " when result coopmat's use is MatrixAccumulatorKHR";
+      }
+      break;
+    default:
+      assert(0 && "a unsupported CoopMat use");
```

Comment:
I think this will be hit if the use is a spec constant so it probably shouldn't assert. Would be good to cover in a test.

Reply:
The CM spec states that the dimensions and the scope parameters may be specialized, but it did not say the matrix use can be specialized. So do you want to have a test case that uses a specialization constant to specify the use parameter and the validator to check that it is not allowed ?

Reply:
The spec says the use is a constant instruction which includes specialization constants. It had to relax the scope because SPIR-V core spec requires that the scope be an OpConstant.